web3/middleware/signing.py

```diff
@@ -0,0 +1,64 @@
+from eth_keys.datatypes import PrivateKey
+
+from web3.utils.datastructures import HexBytes
+# from web3.utils.signing import LocalAccount
+
+import logging
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def construct_transaction_signing_middleware(private_key):
+    """Intercept a transaction in flight and sign it.
+
+    Args:
+        private_key (TYPE): Private key used to sign the transaction.
+                            Accepts the following formats:
+                            - `PrivateKey` object from `eth-keys`
+                            - `LocalAccount` object from `web3.eth.account`
+                            - raw bytes
+                            - TODO: hex encoded string
+
+
+    Returns:
+        TYPE: Description
+    """
+
+    # Check if the private_key is either
+    # - eth_keys.keys.PrivateKey
+    # - web3.eth.account.create().privateKey -- can just check that its HexBytes
+    # - bytes
+    # TODO: - hexbytes
+    valid_key_types = (PrivateKey, HexBytes, bytes)
+    _private_key = None
+
+    for t in valid_key_types:
+        if isinstance(private_key, t):
+            # Might also want to check that bytes length == 32
+            _private_key = private_key
+            break
+
+    def transaction_signing_middleware(make_request, web3):
+        def middleware(method, params):
+            # Only operate on the `eth.sendTransaction` method
+            # Only operate on if the private key matches the public key in the transaction
+            # Note: params == transaction in this case.
+            if method is not 'eth_sendTransaction':
```

Comment:
My python says that `is` is the wrong operator here and that this should be `!=` or `if not thing == other` since the `is` operator is supposed to operate on whether two things are the *same* which is subtly different from equality.

It appears that this does in fact work as expected but I still think it's semantically incorrect.

Reply:
fixed, along with the `nonce` value.

Reply:
Just to point out that this isn't only a hypothetical or style issue, it is pretty easy to bust the string interning* that cpython does:

```
In [80]: 'ab' is 'ab'
Out[80]: True

In [81]: 'ab' is ''.join('ab')
Out[81]: False

In [82]: 'ab' == ''.join('ab')
Out[82]: True
```

*Reuse of instances of string literals.